src/crimson/net/SocketConnection.cc

```diff
@@ -330,14 +330,25 @@ seastar::future<> SocketConnection::keepalive()
 
 seastar::future<> SocketConnection::close()
 {
-  // unregister_conn() drops a reference, so hold another until completion
-  auto cleanup = [conn = ConnectionRef(this)] {};
+  if (state == state_t::closed) {
+    // already closing
+    assert(close_ready.valid());
+    return close_ready.get_future();
+  } else {
```

Comment:
nit: drop the else block to save indentation

Reply:
edit: looking closer at `read_tags_until_next_message()`, it doesn't wait on the future returned by `close()`. maybe we should just change it to only call `close()` if `state != state_t::closed`?

Reply:
> maybe we should just change it to only call close() if state != state_t::closed?

I think not ... `close()` in `read_tags_until_next_message()` can be called first, this means we should also add condition `state != state_t::closed` in `shutdown()`. But then `shutdown()` cannot wait for this `close()`.

Another possible case is that multiple continuations may need to wait for `shutdown()`.

When implementations go complex, IMO we need to provide such interface to allow multiple continuations waiting for `close()`.

Reply:
> multiple continuations may need to wait for `shutdown()`.

could you share with us a use case? i think the connection will only be closed by a single party. hence it will be the one closing the connection who is waiting for this future.

Reply:
No, there is no use case yet (just want to explain the limitations).

The real issue happening now is that the connection can be closed at multiple places (`read_tags_until_next_message()` and `shutdown()`) and **in random orders**. Just checkout the "error_code fix" and run `unittest_seastar_messenger` and there will be assertion error happening:
```
/root/ceph/src/crimson/net/SocketMessenger.cc: In function 'virtual void ceph::net::SocketMessenger::unregister_conn(ceph::net::ConnectionRef)' thread 7f8e90099e40
/root/ceph/src/crimson/net/SocketMessenger.cc: 190: FAILED ceph_assert(found != connections.end())
 ceph version 14.0.0-2679-g2001145 (20011453a24fde224328b352e91643a279396d04) nautilus (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f8e8744f8b3]
 2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f8e8744fa35]
 3: ./unittest_seastar_messenger() [0x4d9b33]
 4: (ceph::net::SocketConnection::close()+0x5e) [0x51351e]
 5: ./unittest_seastar_messenger() [0x4b9243]
 6: ./unittest_seastar_messenger() [0x503558]
 7: (seastar::reactor::run_tasks(seastar::reactor::task_queue&)+0x85) [0x5620f5]
 8: (seastar::reactor::run_some_tasks()+0xe7) [0x562477]
 9: (seastar::reactor::run()+0xcf0) [0x5bdfa0]
 10: (seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&)+0xa54) [0x53ff64]
 11: (seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&)+0xd0) [0x540dd0]
 12: (seastar::app_template::run(int, char**, std::function<seastar::future<> ()>&&)+0xd0) [0x540f50]
 13: (main()+0x602) [0x4bb4f2]
 14: (__libc_start_main()+0xf0) [0x7f8e858cf830]
 15: (_start()+0x29) [0x4bce79]
Aborting on shard 0.
```